### ISSUE TITLE
Proposal: Remove `targetEl`

### ISSUE THREAD

**Issue by LinusBorg** [LinusBorg/portal-vue#74]
### Why

* It's overly complicated for what it does. You can simply do ` new Vue(PortalTarget).$mount(el)`
* ~~It won't work with the upcoming `PortalTargetGroup` component which will (hopefully) allow multiple parallel  sources~~

This is just an idea though, and will be a breaking change, so not possible before `2.0`

### What to do instead?

Mount the PortalTarget yourself:

```javascript
import { PortalTarget } from 'portal-vue '
const target =new Vue({
  el: '#target-el',
  ...PortalTarget,
  propsData: { /* props go here*/ },
})

// or if we also want to set `attributes` for the targets root element:
const targetWrapper = new Vue([
  el: '#target-el',
  components: {  PortalTartget },
  template: '<portal-target class="my-custom-class" name="target"/>' // add any attributes on this element
})
```

That should be easy enough to do yourself in your app during mounting.

### Advantages

* reduces API surface and code size. We could cut a consideable number of lines from `portal` and `utils`, and a few from `portal-target`
* No need to cover edge cases in this lib, like mutiple `<portal>` components trying to mount to the same `target-el`
* More flexible: You can set other Target props like `transition` easily. We would have to extend the API even further to support that with `target-el`, which I don't like.

**Comment by Alfredo-Delgado** [comment]
What would this change look like when using portal-vue as a way to treat components spread across a legacy app page as a single vue application?

e.g.
```html
<div id="vue-mount__page-component"></div>
...server rendered static content...
<div id="vue-portal__component--1"></div>
...server rendered static content...
<div id="vue-portal__component--2"></div>
...server rendered static content...
```

**Comment by LinusBorg** [comment]
Well, I already showed how to manually mount a target. You would simply do that where necessary in your page component.

I can offer a more concrete example if you share hwo you use `target-el` today.

**Comment by Alfredo-Delgado** [comment]
I'm currently using it as a way to migrate legacy server rendered pages into proper single applications with a top-down data flow. What comes out of the server look like my example above. My js entry looks like:

```
new Vue({
  el: document.getElementById('vue-mount__page-component'),
  render(h) {
    return h(PageComponent);
  }
});
```
The template on the page component looks like:
```
<div>
  <portal target-el="#vue-portal__component--1">
    <a-component />
  </portal>
  <portal target-el="#vue-portal__component--2">
    <a-component />
  </portal>
</div>
```

**Comment by oleics** [comment]
> It's overly complicated for what it does.

This feature is the main reason to use portal.

Some code:

```js
    <portal v-for="name in componentNames" v-bind:target-el="'#'+name">
      <component :is="name"></component>
    </portal>
```
Sooo handy if someone wants to enhance already existing html.

Maybe create a new project just for this? Anyone? ;-)

**Comment by LinusBorg** [comment]
I understand that this is useful in this way. However the current implementation is lacking - it's overwriting the created portal on each render.

Implementing some sort of caching could solve this, but that would get messy considering that target-el could be dynamically switched like you show here - further increasing size. 

We could reach into internals of Vue to be able to check if a target-el already has a portal mounted, but I l'd like to aviod that and stick to publicly available APIs.

So I'll think about this some more and do a few experiments when I find the time - maybe we can have both by moving the target-el functionality into some sort of seperate file/"plugin" that people can inlcude on demand?

**Comment by oleics** [comment]
Yes, I think some kind of file/plugin for this would be the way to go here. I do not really need the portal-thing, it's the "give me a way to break out of the vue-tree" thing I/we really need for an application we are currently developing. I have some free time right now, so I'll also put some more thoughts/research into it.

**Comment by OriginalEXE** [comment]
@padcom I believe you have misunderstood the intention here, the only thing that will (probably) be removed is the `targetEl` API, but you can still achieve the same thing it does by following the example given above.

**Comment by oleics** [comment]
@OriginalEXE

> The ability to render outside of the vue application

You can't do that by any of the above examples. You have to create multiple root-elements (that is "multiple apps") within one page, and not one root-element.

**Comment by trusktr** [comment]
Interestingly, the only reason I searched for "vue portal" and found `PortalVue` was to render a component outside the Vue app.

**Comment by trusktr** [comment]
> You can simply do `new Vue(PortalTarget).$mount(el)`

This is not true, it has implications. If you run `new Vue(PortalTarget).$mount(document.body)`, for example, you will completely destroy the `<body>` and re-create a whole new one, therefore any outside code relying on existing elements inside the body will completely broken.

For reference, this is what it takes to do it properly (without considering all edge cases):

```js
mountComponent( Component, document.body )

function mountComponent( Component, container ) {

    // `tmp` will get replaced by the DOM generated from the component
    const tmp = document.createElement( 'tmp' )
    tmp.style.display = 'none'
    container.appendChild( tmp )

    let instance

    if ( derivesFromVue( Component ) ) {
        instance = new Component()
    }
    else {
        instance = new Vue( Component )
    }

    // utterly detroys the tmp element
    instance.$mount( tmp )
}

function derivesFromVue( Component ) {
    return Component.prototype instanceof Vue
}
```

End users should not have to know the internals of component instantiation like this. Many Vue users are using single-file-components, so they will not ever think about these instantiation details.

It would be awesome if you leave this feature in portal-vue, and ensure that it does not destroy outside DOM.

**Comment by trusktr** [comment]
Here's an idea:

Keep the feature, and allow a `keep` option to keep the outside-Vue portal's DOM:

```html
<portal name="" target-el="#widget" keep>...</portal>
```

And in this case just mount the component at the end of the target element.

**Comment by DeyLak** [comment]
I'm also using this library only for that `targetEl` api, cause I think, that the portal feature itself is not super-usefull for SPAs, managed by Vuex or smth like that. Since, you are storing your state in one place, you can build single-way data flow and render something in another component, just relying on your state. I'm afraid, that overuse of portals can make components structure less obvious and incapsulated. 
For example, [react alternative](https://reactjs.org/docs/portals.html) is concentrated on rendering to DOM elements, cause they think that this is the main use of that feature.

**Comment by LinusBorg** [comment]
Seems I have to clesar afew things up.

### The feature won't be deprecated/removed

I change my mind, seeing how valuable it seems to be for most people. I will however rewrite it for the next version to make it a bit more reliable. Wether or not that means a breaking change in the API ist left to be seen.

### Some Notes

@trusktr 

While I agree this most of what you wrote, I have to disagree on this one:

> End users should not have to know the internals of component instantiation like this. Many Vue users are using single-file-components, so they will not ever think about these instantiation details.

That's wrong, as literally *every* user has to call .$mount() at least once, when they mount the main instance of thei app. While in many situations, this mount is hidden behind the `el:` option, which does the mount for them, they, the same caveats still apply: The element that you mount to will be **destroyed and replaced**. 

So that's something Vue developers should *definitely* be aware of. If our docs don't document this clearly enough, they should be adjusted.

@DeyLak I agree that this library's use case for SPA's is limited, but even in SPAs you have challenges like rendering a modal reliably correctly, which usually means moving the code to the verx end of `<body>` - something `portal-vue` is perfect for.

>  I'm afraid, that overuse of portals can make components structure less obvious and incapsulated.

Most things can be overused and doing so usually has a negative effect, so I do agree with the general sentiment.

**Comment by trusktr** [comment]
> That's wrong, as literally every user has to call .$mount() at least once

I've only ever used `new Vue`.

> this mount is hidden behind the el: option

So it means I never really cared about `$mount()` internals. I bet most people haven't cared about it.

> The element that you mount to will be destroyed and replaced.

And most people understand that part and roll with it. But what I mean is that most people won't take the time to learn how to actual mount a component _inside_ another element, they will just stick with whatever Vue does (replace an element). Most users will not want to write a function like the above `mountComponent`, unless they are doing library-ish type stuff.

When I needed to make an adapter that could mount either React or Vue components inside an element, I had to figure this out the hard way, because it isn't entirely clear from the docs (at least no at the time I read them).

> So that's something Vue developers should definitely be aware of

I still disagree for the above reasons. Most people will use `new Vue` and just roll with whatever behavior comes with it not knowing how Vue does it internally.

> If our docs don't document this clearly enough, they should be adjusted.

Yeah, definitely! (if not already)

I would suggest that Vue could even provide (and document) a helper like `mountComponent` so that people still don't have to think about it. Or maybe even an option:

```js
import App from './single-file-component'

new Vue({
  ...App,
  el: '#root',
  replace: false, // will mount inside of #root, instead of replacing.
})
```

With all that said, I'm glad the feature lives!
